api/envoy/config/listener/v3/listener_components.proto

```diff
@@ -112,6 +112,9 @@ message FilterChainMatch {
 
   reserved 1;
 
+  // The Connection Metadata of the connection to match against.
+  string connection_metadata = 100;
```

Comment:
Why do we have a new notion of connection metadata? Doesn't the StreamInfo associated with a connection, providing both dynamic metadata and FilterState suffice?
/wait

Reply:
First of all, thanks @htuch for reviewing the code. I am not familiar with StreamInfo. Are you referring to https://www.envoyproxy.io/docs/envoy/v1.21.0/configuration/http/http_filters/lua_filter.html?highlight=streaminfo#config-http-filters-lua-stream-info-wrapper ? In any case, we will still need the changes in FilterChainMatch to match based in the dynamic metadata correct?

Reply:
FilterChain matching is done based on Network::ConnectionSocket object. 
We are attaching the metadata to Network::ConnectionSocket. 
During filter chain match, this is referenced as below:
```
FilterChainManagerImpl::findFilterChain(const Network::ConnectionSocket& socket) const {
  const auto& connection_metadata = absl::AsciiStrToLower(socket.connectionInfoProvider().connectionMetadata());
  ...
```
Now, streamInfo is on Connection itself and not on the Socket and is referenced as below `connection().streamInfo().setDynamicMetadata() `

How do you propose we use streamInfo to do filterChain match?

Reply:
The caller should have access to this, e..g. in `ActiveStreamListenerBase::newConnection`. @lambdai does this work?

Reply:
Right, the change in this PR kicks in filterChainManager().findFilterChain(*socket) in the code below. So it will still require the socket to hold the metadata and the findFilterChain to select the filter chain based on the metadata.

```
void ActiveStreamListenerBase::newConnection(Network::ConnectionSocketPtr&& socket,
                                             std::unique_ptr<StreamInfo::StreamInfo> stream_info) {
  // Find matching filter chain.
  const auto filter_chain = config_->filterChainManager().findFilterChain(*socket);
 ...
}
```

Reply:
@htuch Are you suggesting this?

We have the streamInfo in the caller:
ActiveStreamListenerBase::newConnection(Network::ConnectionSocketPtr&& socket,
                                             std::unique_ptr<StreamInfo::StreamInfo> stream_info)
                                             
From streaminfo we are anyway extracting the connection_metadata using StreamInfoImpl::downstreamAddressProvider().connectionMetadata()

Inside, newConnection() maybe we can pass this streamInfo as an optional parameter to findFilterChain() ( or overload the findFilterChain method with the additional parameter)?

Reply:
I think metadata is available for listener filters so it must be created before the active connection. My meta-point is that previously when I requested this feature, we decided to revamp filter chain matching to be unified matching instead. I''d be happy if we revisit that decision, but wanted to make sure we don't churn.

Reply:
We have a need for this feature right now @htuch. We are forked off and doing a release of a local branch as a workaround

Reply:
https://github.com/envoyproxy/envoy/pull/18871. In any case, I think we should be using the existing stream info dynamic metadata or filter state whichever way we decide to order these, rather than introduce a third notion of metadata into the API / implementation.

Reply:
I mentioned above that the filterChainMatch function is on Listener Socket. How do you suggest that the existing stream info dynamic metadata made available to the listener socket?
```
void ActiveStreamListenerBase::newConnection(Network::ConnectionSocketPtr&& socket,
                                             std::unique_ptr<StreamInfo::StreamInfo> stream_info) {
  // Find matching filter chain.
  const auto filter_chain = config_->filterChainManager().findFilterChain(*socket);
 ...
}
```

Reply:
Yes, pass it in. @kyessenov can we land #18871 soon?

Reply:
> newConnection (Network::ConnectionSocketPtr&& socket,
                                             std::unique_ptr<StreamInfo::StreamInfo> stream_info) 
  // Find matching filter chain.
> Yes, pass it in

It's unclear to me what is the metadata setter in this PR. Can the setter set metadata in stream info? We can construct the stream info earlier.

However, currently stream_info copies lots of attributes from socket. If we pass both stream info and socket here, it would be very confusing that what is the source of truth: stream_info or socket? What stream info are supposed to be available?

i suggest we don't offer stream_info here. Instead we put it in socket, and eventually listener components copies the metadata to stream info

Reply:
Currently listener filters can set metadata right? This occurs before filter chain match. So, why do we have a special kind of metadata just for filter chain match?

Reply:
Oh, yes, the setter is added to support proxy protocol listener filter 

+1 to define a magical metadata namespace for filter chain match

Reply:
Does this filter metadata come from the ProtoBuffer (via envoy configuration)?  We want to set this metadata through an API. Is this supported?

```
message Metadata {
  // Key is the reverse DNS filter name, e.g. com.acme.widget. The envoy.*
  // namespace is reserved for Envoy's built-in filters.
  map<string, google.protobuf.Struct> filter_metadata = 1;
}
```

Reply:
> This occurs before filter chain match. So, why do we have a special kind of metadata just for filter chain match?

@lambdai - Please point me to the setter you are referring to?

Reply:
> void setDynamicMetadata(const std::string& name, const ProtobufWkt::Struct& value) override {

Yes, this one.

Regarding the key, we can define magical one. 
Your added metadata should be accessible after the filter chain match so we have to name it anyway. Assign a magical key name, e.g. `envoy.connection` and you can migrate to the unified filter chain match framework.

Reply:
@lambdai  - How do you propose we pass in the metadata from here? The StreamInfo is missing in this context.

```
EnvoyQuicProofSource::CertConfigWithFilterChain
EnvoyQuicProofSource::getTlsCertConfigAndFilterChain(const quic::QuicSocketAddress& server_address,
                                                     const quic::QuicSocketAddress& client_address,
                                                     const std::string& hostname) {
  ENVOY_LOG(trace, "Getting cert chain for {}", hostname);
  // TODO(danzh) modify QUICHE to make quic session or ALPN accessible to avoid hard-coded ALPN.
  Network::ConnectionSocketPtr connection_socket = createServerConnectionSocket(
      listen_socket_.ioHandle(), server_address, client_address, hostname, "h3");
  const Network::FilterChain* filter_chain =
      filter_chain_manager_->findFilterChain(*connection_socket);
...
```

Reply:
Do you want to fulfill the new findFilterChain signature, or do you want to set connection metadata in quic ? 
I believe it's the former because quic doesn't formulate a `connection` at this moment.

It's likely we should pass a dumb stream info here similar to the return value of  `createServerConnectionSocket`

Reply:
We are planning to work on this. Leaving a comment so PR does not close

Reply:
Unified matcher has been merged https://github.com/envoyproxy/envoy/pull/20110. We still need a metadata matcher but that would be an extension now.

Reply:
@kyessenov - This is great to hear. Could you please share the design for Unified matcher and how to integrate metadata matcher to unified filter?

Reply:
Here's an overview https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/advanced/matching/matching_listener#arch-overview-matching-listener.

Reply:
Hey @kyessenov - we are unable to use the unified matcher as it doesn't support custom matcher we are building for dynamic metadata matching https://github.com/envoyproxy/envoy/issues/23503

Reply:
@kyessenov - Is there a way to support dynamic metadata matching in addition to static string matching in Unified matcher?